content/news/XXXX-XX-XX-dvs-internals-pt1.md

```diff
@@ -0,0 +1,157 @@
+
+You may wonder why the timecode signal in the screenshots looks like a analog signal instead of a "lollypop chart".
+*Digital* vinyl scratch system work with digital signals, not analog ones, right?
+And you'd be right, it's just that most audio tools (like Audacity) show a wave instead of individual samples by default.
+
+## Detecting Zero Crossings
+
+Now that we know the basics of digital audio, we can start thinking about a simple low-latency approach to detect the playback velocity using so-called "zero crossings".
+
+Whenever the wave goes from the positive part to the negative part or vice-versa, it crosses "zero".
+For example, in the timecode signal screenshot I used above, we can see 3 zero crossings:
+
+![Zero crossings in timecode signal]({static}/images/news/timecode-signal-samples.png)
+
+Note that we're working with a digital signal.
+To illustrate that this signal is time-discrete, each individual sample value is illustrated by a little dot.
+
+To detect these zero crossings reliably, we can't just check if the sample value equals zero, because we cannot assume that a sample measurement is happening at the exact time that the signal is at zero position.
+
+It might be the case the current sample is *after* the zero position and the previous sample was *before* the zero position.
+Hence, we also compare the previous sample value with the current sample value.
+If the current value is positive and the previous value was negative, or if the previous value was negative and the current value is positive, the signal crossed zero.
+
+## Refining Pitch Detection Using Zero Crossings
+
+Now that we know how to detect zero crossings, we can use them for determining the pitch.
+
+If we look at the timecode signal, we can see that each cycle of the timecode signal has two zero crossings:
+At the start of the cycle, when the positive part of the wave starts and in the middle of the cycle when the negative part of the wave starts.
+
+So the equivalent of checking for 1000 cycles per second is to check if there are 2000 zero crossings per second.
+Let's say the audio interface uses a sample rate of 44100 Hz, then there should be a zero crossing every 441000 / 2000 = 22.05 samples.
+
+We can now detect the pitch by comparing the expected number of samples between zero crossings with the actual number of samples between them.
+This very simple, low latency pitch detection algorithm works because the Serato timecode has a fixed frequency of 1000 Hz, which means that the distance between all zero crossing is fixed.
+It would not work with a regular music signal, which has lots of different frequencies added together.
+
+However, we need to use multiple sample distances and calculate the average to make this calculation more accurate, because in reality there are no fractions of samples, only full samples.
```

Comment:
I don't understand that sentence, can you elaborate?

Reply:
To calculate the actual sample distance (which may be fractional), you need more than one measurement, because it's not possible measure factional sample distances (because sampling rate = how often do you measure the signal per second).

Let's say the distance between two zero crossings is 99.9 samples. But the actual measurement will be 100 samples in 9 out of 10 cases and 99 samples in 1 out of 10 cases.

Do you have a suggestion how to explain it better?

Reply:
I still don't understand what exactly a "sample distance" is supposed to be. Do you mean "the distance of samples between adjacent zero crossings"?

Reply:
Yes.

Reply:
Ah yes, sorry its getting late. ;)

> Do you have a suggestion how to explain it better?

Not really, maybe tomorrow.